hapi-fhir-base/src/main/java/ca/uhn/fhir/util/ResourceUtil.java

```diff
@@ -79,4 +164,499 @@ public static String getRawStringFromResourceOrNull(@Nonnull IBaseResource theRe
+
+		for (IBase fromFieldValue : filteredFromFieldValues) {
+			IBase newFieldValue;
+			if (Strings.CI.equals(fromFieldValue.fhirType(), "CodeableConcept")) {
+				newFieldValue = mergeOrClone(
+						theFhirContext,
+						theChildDefinition,
+						fromFieldValue,
+						theTargetFieldValues,
+						theMergeControlParameters,
+						targetValue -> isCodeableConceptMergeCandidate(
+								fromFieldValue, targetValue, terser, theMergeControlParameters));
+			} else if (Strings.CI.equals(fromFieldValue.fhirType(), "Coding")) {
+				newFieldValue = mergeOrClone(
+						theFhirContext,
+						theChildDefinition,
+						fromFieldValue,
+						theTargetFieldValues,
+						theMergeControlParameters,
+						targetValue ->
+								isCodingMergeCandidate(terser, fromFieldValue, targetValue, theMergeControlParameters));
+			} else {
+				newFieldValue = createNewElement(terser, theChildDefinition, fromFieldValue);
+			}
+
+			if (newFieldValue != null) {
+				try {
+					theTargetFieldValues.add(newFieldValue);
+				} catch (UnsupportedOperationException e) {
+					theChildDefinition.getMutator().setValue(theTarget, newFieldValue);
+					theTargetFieldValues = theChildDefinition.getAccessor().getValues(theTarget);
+				}
+			}
+		}
+	}
+
+	/**
+	 * If <code>theMergePredicate</code> identifies one of the elements in <code>theTargetFieldValues</code>
+	 * as a match for <code>theSourceFieldValue</code>, those two elements will be merged.
+	 * Otherwise, <code>theSourceFieldValue</code> will be cloned and returned.
+	 * @param theFhirContext            Context holding resource definition
+	 * @param theChildDefinition        The definition of the field being merged
+	 * @param theSourceFieldValue       A value from the source resource
+	 * @param theTargetFieldValues      The values already present in the target resource
+	 * @param theMergeControlParameters Parameters to provide fine-grained control over the behaviour of the merge
+	 * @param theMergePredicate         An algorithm to identify matching resources
+	 * @return a clone of <code>theSourceFieldValue</code> if none of the values in <code>theTargetFieldValues</code>
+	 *         is a match. Otherwise, null, indicating that a merge took place
+	 */
+	private static IBase mergeOrClone(
+			FhirContext theFhirContext,
+			BaseRuntimeChildDefinition theChildDefinition,
+			IBase theSourceFieldValue,
+			List<IBase> theTargetFieldValues,
+			MergeControlParameters theMergeControlParameters,
+			Predicate<IBase> theMergePredicate) {
+		IBase newFieldValue = null;
+		FhirTerser terser = theFhirContext.newTerser();
+		Optional<IBase> matchedTargetValue =
+				theTargetFieldValues.stream().filter(theMergePredicate).findFirst();
+		if (matchedTargetValue.isPresent()) {
+			mergeAllFields(theFhirContext, theSourceFieldValue, matchedTargetValue.get(), theMergeControlParameters);
+		} else {
+			newFieldValue = createNewElement(terser, theChildDefinition, theSourceFieldValue);
+		}
+		return newFieldValue;
```

Comment:
Minor suggestion (can be left as is), returning a value in each branch reads a bit better.

```suggestion
		FhirTerser terser = theFhirContext.newTerser();
		Optional<IBase> matchedTargetValue =
				theTargetFieldValues.stream().filter(theMergePredicate).findFirst();
		if (matchedTargetValue.isPresent()) {
			mergeAllFields(theFhirContext, theSourceFieldValue, matchedTargetValue.get(), theMergeControlParameters);
			return null;
		} else {
			return createNewElement(terser, theChildDefinition, theSourceFieldValue);
		}
```